src/java.base/share/classes/java/lang/invoke/VarHandle.java

```diff
@@ -2016,15 +2016,35 @@ static final class AccessDescriptor {
         final Class<?> returnType;
         final int type;
         final int mode;
+        // The cache for adapted MH if the access site has a constant VH
+        // that does not match the erased type for.
+        @Stable MethodHandle adaptedMh;
 
-        public AccessDescriptor(MethodType symbolicMethodType, int type, int mode) {
+        AccessDescriptor(MethodType symbolicMethodType, int type, int mode) {
             this.symbolicMethodTypeExact = symbolicMethodType;
             this.symbolicMethodTypeErased = symbolicMethodType.erase();
             this.symbolicMethodTypeInvoker = symbolicMethodType.insertParameterTypes(0, VarHandle.class);
             this.returnType = symbolicMethodType.returnType();
             this.type = type;
             this.mode = mode;
         }
+
+        @ForceInline
+        MethodHandle adaptedMethodHandle(VarHandle vh) {
+            var constant = MethodHandleImpl.isCompileConstant(vh);
+            var cache = adaptedMh;
+            if (constant == MethodHandleImpl.CONSTANT_YES && cache != null) {
+                return cache;
+            }
+
+            // This is still a hot path if vh is not constant - in this case,
+            // asType is the bottleneck for constant folding, unfortunately
+            var result = vh.getMethodHandle(mode).asType(symbolicMethodTypeInvoker);
```

Comment:
`mode` and `symbolicMethodTypeInvoker` are part of `AccessDescriptor` while `vh` comes as an argument. What guarantees that a cached adapter is compatible with `vh` observed during subsequent calls? It means that `vh` shape stays exactly the same shape. Is it correct? Would be good to have it validated with asserts.

Reply:
I am assuming that the previous `vh` observed is compatible with future ones if compiler can fold the `vh` into a constant. If it is not, we can drop the updates to the cache field in the C2 compiled slow code.